Fix Failure to merge input DEX and driver DEX

mohe2015 opened the pull request:
This SHOULD work but I only tried it on my device.

Closes #17

mohe2015 commented on driver/src/main/java/org/cf/oracle/Driver.java:
```diff
@@ -19,7 +19,7 @@
 public class Driver {
 
-    private static final String DRIVER_DIR = "/data/local";
+    private static final String DRIVER_DIR = "/data/local/tmp";
```
WHYTHISWORKS: On my phone this is the correct temp dir

CalebFenton replied:
On my phone, it's not the correct temp dir. :D

A better solution would be to determine the best directory.

mohe2015 commented on driver/src/main/java/org/cf/oracle/Driver.java:
```diff
@@ -77,6 +77,13 @@ public static void main(String[] args) {
+            die("Failed to load native libraries", e);
+        }
+
+        try {
```
WHYTHISWORKS: does nothing, could be removed

CalebFenton replied:
Rather than hard coding this, how about making it a command line option?

mohe2015 commented on lib/dex-oracle/driver.rb:
```diff
@@ -47,38 +47,18 @@ def install(dex)
-        exit -1
-      end
-      tf.close
-
```
WHYTHISWORKS: This produced the following errors:
`[ERROR] 2017-08-10 23:06:38: Failure to merge input DEX and driver DEX:
Exception in thread "main" com.android.dex.DexIndexOverflowException: method ID not in [0, 0xffff]: 65536
	at com.android.dx.merge.DexMerger$6.updateIndex(DexMerger.java:495)
	at com.android.dx.merge.DexMerger$IdMerger.mergeSorted(DexMerger.java:261)
	at com.android.dx.merge.DexMerger.mergeMethodIds(DexMerger.java:504)
	at com.android.dx.merge.DexMerger.mergeDexes(DexMerger.java:161)
	at com.android.dx.merge.DexMerger.merge(DexMerger.java:188)
	at com.android.dx.merge.DexMerger.main(DexMerger.java:1120)


[ERROR] 2017-08-10 23:06:38: Your input DEX inexplicably contains const-string and const-string/jumbo. This probably means someone fucked with it. In any case, it means DexMerge is failing because there are too many strings.
Try this: baksmali the DEX, replace all const-string instructions with const-string/jumbo, then recompile with smali and use that DEX as input. Sorry, I don't want to do this for you. It's too complicated.`

The second one is simply an invalid one and I think the first one is because of too many methods but I'm not sure

mohe2015 commented on lib/dex-oracle/driver.rb:
```diff
@@ -47,38 +47,18 @@ def install(dex)
       tempzip_path = tz.path
       tz.close
-      Utility.create_zip(tempzip_path, 'classes.dex' => tf)
+      Utility.create_zip(tempzip_path, 'classes.dex' => Resources.driver_dex, 'classes2.dex' => dex)
```
WHYTHISWORKS: This adds the dex file of the apk and the dex file of the driver into a zip file which then gets uploaded to the device. The driver loads all classes from that zip file so it also finds the ones from the apk.

CalebFenton replied:
Does _Android_ load classes2.dex automatically? I can't test it myself right now and I don't see the code to load it.

mohe2015 replied:
I think app_process does that because as far as I know normal apks can also contain multiple dex files which are named the same way. But you should check that before merging.

mohe2015 replied:
Please check + verify my implementation. I had to use an android project as the java project could not use DexClassLoader.

mohe2015 commented on lib/dex-oracle/driver.rb:
```diff
@@ -155,6 +135,8 @@ def get_driver_dir
       next if stdout == "ls: #{dir}: Permission denied"
       next if stdout == "ls: #{dir}: No such file or directory"
+      stdout = adb("shell ls #{dir}")
+      next if stdout == "opendir failed, Permission denied"
```
WHYTHISWORKS: For me shell -x did not work so I simply added what worked for me.

CalebFenton replied:
I don't remember what shell -x does and I'm hesitant to remove it without understanding.

mohe2015 commented:
I added comments directly into the diff of the pull request: https://github.com/CalebFenton/dex-oracle/pull/16/files

I tried the tool with Instagram (3cfca728a62dcf53d396cda8d8805bd856fa6d63) because I didn't have anything better.

mohe2015 commented:
ping (maybe a little bit too early) @CalebFenton

CalebFenton commented on lib/dex-oracle/driver.rb:
```diff
@@ -47,38 +47,18 @@ def install(dex)
       tempzip_path = tz.path
       tz.close
-      Utility.create_zip(tempzip_path, 'classes.dex' => tf)
+      Utility.create_zip(tempzip_path, 'classes.dex' => Resources.driver_dex, 'classes2.dex' => dex)
```
This is how it should be done. Trying to merge dex files was "cool" back when people used fewer than 65535 methods.

I noticed the driver.dex changed but I don't see any changes to the driver Java code.

CalebFenton commented:
Thanks for the reminder. :D

mohe2015 commented:
I added code to get the correct temp directory. Merge? Maybe I (or somebody else) will add the command line option to load native libraries later.

CalebFenton commented on driver/src/main/java/org/cf/oracle/Driver.java:
```diff
@@ -92,13 +91,33 @@ private static void showUsage() {
 
     public static void main(String[] args) {
+        // find out the driver dir as it is different on some devices
+        for (String driverDir : new String[] { "/data/local/tmp", "/data/local" }) {
```
Why not just assume files are in the same directory? Or get current working directory? It's cleaner than hard coding temp paths in two places.

CalebFenton commented:
Tomorrow or this weekend I should have enough free time to test this (especially the `shell -x` thing) in a variety of emulators and hopefully merge it.

CalebFenton commented on spec/dex-oracle/driver_spec.rb:
```diff
@@ -20,7 +20,7 @@
     Driver.new(device_id)
   end
-  let(:driver_stub) { 'export CLASSPATH=/data/local/od.zip; app_process /system/bin org.cf.oracle.Driver' }
+  let(:driver_stub) { 'cd /data/local; export CLASSPATH=/data/local/od.zip; app_process /system/bin org.cf.oracle.Driver' }
```
Is this `cd` necessary? Why did it work before?

mohe2015 replied:
I think it is necessary as `org.cf.oracle.Driver` does not have a hard-coded `DRIVER_DIR` anymore and it needs to find it out somehow else.

CalebFenton commented on lib/dex-oracle/driver.rb:
```diff
@@ -210,7 +200,14 @@ def validate_output(full_cmd, full_output)
+      unless exception.end_with?('No such file or directory')
+        adb("shell rm #{@driver_dir}/od-exception.txt")
+        raise exception
+      end
```
This looks like it misses the condition where you have an unexpected header and there's no od-exception.txt, i.e. a catastrophic failure.

CalebFenton commented:
There's a lot going on here and I'd like to know what motivated the changes. For example, if it was working before on your machine, why add the dependency on dexclassloader and pull in the entire android build system? Could you please summarize all the changes?

mohe2015 commented:
Summary of all changes:

- Why I added the `cd #{@driver_dir}`: My driver directory was not the hardcoded one in [Driver.java, line 28](https://github.com/CalebFenton/dex-oracle/pull/16/files#diff-2c2f2c46551723329cbece98348ae254L28), so I needed to determine it somehow. As duplicated code is bad I thought the best idea would be to just set it as working directory.


- Why I added the DexClassLoader:
 As you already pointed out and according to https://developer.android.com/studio/build/multidex.html, merging dex files is not always possible. Therefore my first idea was to just put both dex files in the zip file.
This also works for Android 5.0 or higher: 
   > Android 5.0 (API level 21) and higher uses a runtime called ART which natively supports loading multiple DEX files from APK files. ART performs pre-compilation at app install time which scans for classesN.dex files and compiles them into a single .oat file for execution by the Android device. Therefore, if your minSdkVersion is 21 or higher, you do not need the multidex support library. (https://developer.android.com/studio/build/multidex.html).

    But you also need to support devices below sdk version 21 so I needed to load the dex manually. As the DexClassLoader can only load classes.dex and not classes1.dex, I simply created a new zip file which gets pushed to the device and loaded by [TargetParser.java](https://github.com/CalebFenton/dex-oracle/pull/16/files#diff-a5cd7bc3162afb6196623c22f7f57c8aR112). The problem was that the DexClassLoader class only exists in the android build system so I had to pull it in.

CalebFenton commented:
Thanks for the explanations. I didn't know android 5 did that. Cool.

I don't like adding all of the complexity of an android project. In my experience, they're a lot of headache. It seems like it's not need because older apps should still run on android 5. If Dex Oracle just required android 5 or higher emulator, everything should be fine. Is this wrong?

I think you could also just build against the android framework stub jar if you really need DexClassLoader.

mohe2015 commented:
Yes, that would work. I will revert all unnecessary changes.

mohe2015 commented:
Moved to https://github.com/CalebFenton/dex-oracle/pull/21
